Improve outdated version check

Toffikk opened the pull request:
This pull request improves the outdated build check by querying the api if an update is *actually available* instead of blindly logging that you haven't updated in a while. 
And adapts the `PaperVersionFetcher.java` class to the new `fill v3` api with an `User-Agent` field in the format of
- `Paper/1.21.8-DEV-4e41aad (https://papermc.io)` 

When the server finishes loading and prints out `Done`, we query the api asynchronously and return the build status based on the four types below: 
- outdated build & version
<img width="1061" height="40" alt="ultra outdated" src="https://github.com/user-attachments/assets/e5f7feb3-928d-4b77-9b6d-610294e0796c" />

- outdated build
<img width="1082" height="39" alt="image" src="https://github.com/user-attachments/assets/906f3f5c-2b6f-41d5-bad5-e688b49037b8" />

- new version available but latest build (for the outdated version)
<img width="949" height="119" alt="outdated but less" src="https://github.com/user-attachments/assets/afec8203-dda2-4799-acfd-af92a32f1985" />

- up-to-date/the newer version is experimental -> we dont log at all

This check can be disabled by specifying the appropriate system property `-Dpaper.disableStartupVersionCheck`

Doc94 commented on paper-server/patches/sources/net/minecraft/server/MinecraftServer.java.patch:
```diff
@@ -769,6 +769,51 @@
++                t.printStackTrace();
++            }
++        });
++            // Paper end - improve outdated version check
```
i feel this can be moved to a method in PaperVersionFetcher because you use things like urls already defined in that class

Toffikk replied:
Yeah you're right will move this

Toffikk replied:
moved

Doc94 commented on paper-server/src/main/java/org/bukkit/craftbukkit/Main.java:
```diff
@@ -17,6 +17,7 @@ public class Main {
     public static final java.time.Instant BOOT_TIME = java.time.Instant.now(); // Paper - track initial start time
     public static boolean useJline = true;
     public static boolean useConsole = true;
+    public static boolean printedWarning = false;
```
this can have a better name.. maybe "checkVersionOnServerStart" or not sure... but a thing most self-explaining than printWarning

Toffikk replied:
I feel like `printedWarning` is good enough, it's just used as a ref if the later logic should run

Toffikk committed:
Improve outdated version checking

DenWav commented on paper-server/src/main/java/org/bukkit/craftbukkit/Main.java:
```diff
@@ -224,15 +230,20 @@ public static void main(String[] args) {
-                        System.err.println("*** Warning, you've not updated in a while! ***");
-                        System.err.println("*** Please download a new build from https://papermc.io/downloads/paper ***");
-                        // Paper end
+                        System.err.println("*** It seems you have not updated Paper in a while ***");
```
This message really isn't useful at all. Just remove it in my opinion. The message that says if the server is _actually_ out of date is the only useful one.

Toffikk replied:
should we make it check for updates on every startup in that case?

Toffikk replied:
addressed

Toffikk committed:
Address review

Toffikk committed:
Formatting and allow disabling the update checking

powercasgamer commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -26,16 +26,21 @@
 public class PaperVersionFetcher implements VersionFetcher {
     private static final Logger LOGGER = LogUtils.getClassLogger();
+    private static final org.apache.logging.log4j.Logger SIMPLE_LOGGER = org.apache.logging.log4j.LogManager.getRootLogger();
+    private static final net.kyori.adventure.text.logger.slf4j.ComponentLogger COMPONENT_LOGGER = net.kyori.adventure.text.logger.slf4j.ComponentLogger.logger(SIMPLE_LOGGER.getName());
```
Why do you need 3 different loggers here..?

Toffikk replied:
the SIMPLE_LOGGER is used to avoid printing [PaperVersionFetcher] and the COMPONENT_LOGGER is for avoiding + formatted component messages; can probably put both under the COMPONENT_LOGGER

Toffikk replied:
I replaced the SIMPLE_LOGGER with COMPONENT_LOGGER

Toffikk committed:
remove SIMPLE_LOGGER

Y2Kwastaken commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -49,13 +54,69 @@ public Component getVersionMessage() {
+                    } else {
+                        COMPONENT_LOGGER.info(text("*************************************************************************************", NamedTextColor.GREEN));
+                        COMPONENT_LOGGER.info(text("You are running the latest build for your Minecraft version (" + build.minecraftVersionName() + ")", NamedTextColor.GREEN));
+                        COMPONENT_LOGGER.info(text("However, there is a new Minecraft version available on the downloads page (" + newVersion + ")!", NamedTextColor.GREEN));
```
"However" Can be dropped here. It doesn't particularly add to the log statement

Toffikk replied:
I feel like However should stay here, it makes the information flow better

Y2Kwastaken commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -49,13 +54,69 @@ public Component getVersionMessage() {
+                        COMPONENT_LOGGER.info(text("*************************************************************************************", NamedTextColor.GREEN));
+                        COMPONENT_LOGGER.info(text("You are running the latest build for your Minecraft version (" + build.minecraftVersionName() + ")", NamedTextColor.GREEN));
+                        COMPONENT_LOGGER.info(text("However, there is a new Minecraft version available on the downloads page (" + newVersion + ")!", NamedTextColor.GREEN));
+                        COMPONENT_LOGGER.info(text("It is recommended that you download it as soon as possible (unless it is experimental)", NamedTextColor.GREEN));
```
I think this line could be confusing. I'm also not entirely sure about the reminder in general. If someone is using an up-to-date stable version like 1.21.4, it might be annoying to receive warnings about newer versions that are still experimental, like 1.21.5 or 1.21.6.

I'd personally suggesst:
A. Implement a way to detect and exclude experimental versions from triggering this message, or
B. Consider removing this else block entirely.

Toffikk replied:
B is defo out of question since i was asked to have a message of some sort if theres a new minecraft release available, however as for A there is no endpoint (yet) to see if the version is experimental or not but there is a way that would involve grabbing the latest build for that ver to check if its experimental so im going to try to do that

The-Epic replied:
there is a way to check if its experimental, `projects/paper/versions/<ver>/builds` provides a list of builds and what channel they are on

Y2Kwastaken replied:
Considering you were asked to I'd prefer a better solution to check if its experimental then, ~~considering there is no endpoint not much you can do here~~. My personal opinion is that these messages should be direct.
e.g. 
good: "You should update to the new minecraft version" 
bad: "You should update to the new minecraft version, unless the build is experimental"
The former cuts down on any possible confusion that could arise.

Y2Kwastaken commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -49,13 +54,69 @@ public Component getVersionMessage() {
+                    } else {
+                        COMPONENT_LOGGER.error(text("*** Currently you are " + distance + " build(s) behind ***"));
+                        COMPONENT_LOGGER.error(text("*** It is highly recommended to download a new build from " + DOWNLOAD_PAGE + " ***"));
+                        COMPONENT_LOGGER.error(text("*** Also note that a new Minecraft version has been released (" + newVersion + ")! ***"));
```
"Also note that a new" can be dropped as well a simple "A new Minecraft version has been released", though I do have the same problem with this as above considering experimental releases etc.

Toffikk replied:
Will redo the wording on this one

Y2Kwastaken commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -49,13 +54,69 @@ public Component getVersionMessage() {
+                case DISTANCE_ERROR -> COMPONENT_LOGGER.error(text("*** Error obtaining version information! Cannot fetch version info ***"));
+                case 0 -> {
+                    if (!newVersionAvailable) {
+                        COMPONENT_LOGGER.info(text("*** You are running the latest version! ***", NamedTextColor.GREEN));
```
Is it really necessary to log if someone is on the latest version?

Toffikk replied:
Might be nice to let them know but also unnecessary

Y2Kwastaken reviewed (commented):
My comments are a bit nitpicky, so I just wanted to clarify the reasoning behind them. First off, this is a great change, the old two week notice definitely feels due for an update considering that Paper provides a proper downloads API.

The language suggestions are mainly with accessibility in mind. For users who may not be fluent in English. Keeping logging messages concise and free of filler words can really help keep the message is clear and easy to understand, even in translation.

Also important to note I'm not a maintainer just trying to open up some discussion here and suggest some things I see that could be improved.

Toffikk committed:
Address yet another set of reviews

Y2Kwastaken commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -84,33 +84,32 @@ public static void getUpdateStatusStartupMessage(final String repo, final Server
                         COMPONENT_LOGGER.info(text("*************************************************************************************", NamedTextColor.GREEN));
                         COMPONENT_LOGGER.info(text("You are running the latest build for your Minecraft version (" + build.minecraftVersionName() + ")", NamedTextColor.GREEN));
                         COMPONENT_LOGGER.info(text("However, there is a new Minecraft version available on the downloads page (" + newVersion + ")!", NamedTextColor.GREEN));
-                        COMPONENT_LOGGER.info(text("It is recommended that you download it as soon as possible (unless it is experimental)", NamedTextColor.GREEN));
```
You can add the (unless it is experimental) back unless you genuinely plan to check if the version is experimental or not

Toffikk replied:
Refer to below

Y2Kwastaken commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -84,33 +84,32 @@ public static void getUpdateStatusStartupMessage(final String repo, final Server
-                    COMPONENT_LOGGER.error(text("*** You are " + distance + " builds behind! ***"));
-                    COMPONENT_LOGGER.error(text("*** Please download a new build from " + DOWNLOAD_PAGE + " ***"));
-                    if (newVersionAvailable) {
-                        COMPONENT_LOGGER.error(text("*** Also note that a new Minecraft version has been released (" + newVersion + ")! ***"));
```
even this where you're saying hey a new version is available, that is fine unless you are ensuring there is a non experimental upgrade.

Toffikk replied:
If you check i added handling to the method which checks for newer versions which ensures the version has to have a stable build

Y2Kwastaken reviewed (commented):
I mean a lot of your diff is changing what was really fine. My main point is that you should check for experimental instead of just blindly recommending and upgrade. I'd argue a lot of this could be reverted. Explicitly saying hey a new stable Minecraft version is available is fine, what I personally think is not fine is nagging for experimental releases.

Y2Kwastaken commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -49,13 +54,68 @@ public Component getVersionMessage() {
+                case DISTANCE_UNKNOWN -> COMPONENT_LOGGER.warn(text("*** You are running an unknown version! Cannot fetch version info ***"));
+                case 5 -> {
+                    if (!newVersionAvailable) {
+                        COMPONENT_LOGGER.error(text("*** You are " + distance + " builds behind! ***"));
```
This and line 108 have two different ways of saying the same thing might be worth joining them to be more similar. This is more of another question is there any reason a distance of 5 has a separate branch compared to all other distances?

Toffikk replied:
I chose 5 builds as a middleground between "it's recommended to update" and "update asap" but that number can be changed or reverted

Y2Kwastaken replied:
for context that sends a message for only 5 builds out of date, as soon as you hit 6 builds out of date the default is hit again. I think it might be smart to just merge the 2 branches. If people agree its beneficial to have a change of wording at all after 5 builds or so that should be done within the body of the default branch.

electronicboy replied:
It should just tell you how many builds outdated you are, trying to provide recommendations based on # of builds is somewhat silly; numbers mean nothing when the build after the one you're on fixes some high risk crash

Toffikk replied:
oh yeah... gonna have to fix that xD

Toffikk replied:
addressed

Toffikk replied:
> It should just tell you how many builds outdated you are, trying to provide recommendations based on # of builds is somewhat silly; numbers mean nothing when the build after the one you're on fixes some high risk crash

github hid this from me, guess i need to make a yet another commit

Toffikk replied:
should be good now

Toffikk committed:
>=5 not case 5

Toffikk committed:
remove the 5 builds logic

Toffikk committed:
Merge branch 'main' of https://github.com/PaperMC/Paper into better-outdated-version-checking

MiniDigger commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -83,6 +134,57 @@ private static Component getUpdateStatusMessage(final String repo, final ServerB
+    private static @Nullable String fetchMinecraftVersionList(final ServerBuildInfo build) {
+        try {
+            try (final BufferedReader reader = Resources.asCharSource(
+                URI.create("https://api.papermc.io/v2/projects/paper").toURL(),
```
needs to set proper user agent header, i.e. Paper/1.21.5 (https://papermc.io)

Toffikk replied:
done 👍

MiniDigger commented on paper-server/patches/sources/net/minecraft/server/MinecraftServer.java.patch:
```diff
@@ -779,6 +779,15 @@
 +            }
 +            // Paper end - Add onboarding message for initial server start
++            // Paper start - improve outdated version check
++            if (System.getProperty("IReallyKnowWhatIAmDoingISwear") == null) {
```
the system proper changed its meaning and thus should be named better now, something like paper.disableStartupVersionCheck

Toffikk replied:
also done

Toffikk committed:
Merge remote-tracking branch 'papermc/main' into better-outdated-version-checking

Also address reviews

Toffikk committed:
Migrate the class to use Fill

Toffikk committed:
Add finals

Toffikk committed:
Use latest build endpoint

Toffikk committed:
Simplify logic for build walking

the new api response doesnt flip it around

Toffikk committed:
Merge branch 'main' of https://github.com/PaperMC/Paper into better-outdated-version-checking

Toffikk committed:
Merge branch 'main' of https://github.com/PaperMC/Paper into better-outdated-version-checking

Toffikk committed:
Merge branch 'main' of https://github.com/PaperMC/Paper into better-outdated-version-checking

Owen1212055 commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -83,12 +136,72 @@ private static Component getUpdateStatusMessage(final String repo, final ServerB
 
+    private static @Nullable String fetchMinecraftVersionList(final ServerBuildInfo build) {
+        final String currentVersion = build.minecraftVersionId();
+        final String userAgent = "Paper/" + currentVersion + " (https://papermc.io/)";
```
In the user agents, can you please use VERSION_SIMPLE rather than just minecraft version?
Additionally, use the brand name rather than hard coded "Paper"

Toffikk replied:
adjusted ^^

Owen1212055 commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -83,12 +136,72 @@ private static Component getUpdateStatusMessage(final String repo, final ServerB
-            ).openBufferedStream()) {
+            final URL buildsUrl = URI.create("https://fill.papermc.io/v3/projects/paper/versions/" + build.minecraftVersionId()).toURL();
+            final HttpURLConnection connection = (HttpURLConnection) buildsUrl.openConnection();
+            connection.setRequestProperty("User-Agent", userAgent);
```
here too

Toffikk committed:
Switch to using VERSION_SIMPLE and brand name

Toffikk committed:
Capitalize `Minecraft`

Toffikk committed:
Merge remote-tracking branch 'papermc/main' into better-outdated-version-checking

Toffikk committed:
cleanup logic in version fetcher

Toffikk committed:
fix indent

Toffikk committed:
more cleanup

Toffikk committed:
use stream instead of loop

Toffikk committed:
build(s)

Leguan16 commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -46,22 +51,68 @@ public long getCacheTime() {
+                    COMPONENT_LOGGER.info(text("However, there is a new Minecraft version available on the downloads page (" + newRelease + ")!", NamedTextColor.GREEN));
+                    COMPONENT_LOGGER.info(text("It is recommended that you download it as soon as possible", NamedTextColor.GREEN));
+                    COMPONENT_LOGGER.info(text(DOWNLOAD_PAGE, NamedTextColor.GREEN));
+                    COMPONENT_LOGGER.info(text("*************************************************************************************", NamedTextColor.GREEN));
```
I don't know if I am a fan of the green coloring here. Maybe just make it a warn? And then I don't think a `ComponentLogger` is needed at all.

Toffikk replied:
I thought about making that a warn aswell but that would make it quite “imposing” for server owners running older versions on purpose, but then again they shouldn’t run outdated versions… i’m open to feedback on this

Leguan16 replied:
If they want to run it on purpose they can just use the system property to disable the check.

Toffikk replied:
should we abandon the use of root logger in this case too? and have the class name printed next to the info or keep it as it is right now with the root logger and cleaner output

Toffikk replied:
Made this a warn but kept the component logger for now

Toffikk committed:
Merge remote-tracking branch 'papermc/main' into better-outdated-version-checking

Toffikk committed:
Rebase for 1.21.10

Squashed commit of the following:

commit c33a9ce18cf4f988f80fb23851003f4e5e8a421a
Author: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date:   Mon Oct 13 14:10:41 2025 -0700

    Fix incorrect variable use in Entity#startRiding

    This diff applied and compiled after updating Parchment, but was incorrectly pointing to a field instead of the parameter

    fixes #13182

commit ffcb7b22092c2838498f8a0113e14996a15af2b0
Author: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date:   Mon Oct 13 21:55:54 2025 +0200

    Update Parchment (#13177)

    Co-authored-by: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
    Co-authored-by: Pedro <3602279+Doc94@users.noreply.github.com>

commit 610f1d2177f9e6320503e3dad6c42a93a6e48d56
Author: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date:   Sat Oct 11 18:30:55 2025 -0700

    Update fill-gradle to v1.0.9

commit 9d95cd55a883414b2f46e8d6b1fcd70de3bc6a9b
Author: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date:   Sat Oct 11 17:52:08 2025 -0700

    Use BUILD_STARTED_AT instead of Instant.now() for build timestamp (#13175)

commit ce983d73f0111065337c8fdd4d1b713bbc8d41cf
Author: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date:   Sat Oct 11 16:49:08 2025 -0700

    Misc fixes to tick reporting (#13174)

commit ba2fb8c9a3629b28fd01511f8670db339173603b
Author: lucko <git@lucko.me>
Date:   Sat Oct 11 14:56:31 2025 +0100

    Update spark-paper dependency version (#13171)

commit a0ea7297c52bd5aee889c5c1b101c7d593c049b0
Author: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date:   Fri Oct 10 19:19:55 2025 -0700

    Fix minimum tick time reporting and off thread reading

commit 056268edf2e106482e7c2748be08a85ec465c298
Author: JustAHuman-xD <65748158+JustAHuman-xD@users.noreply.github.com>
Date:   Fri Oct 10 19:19:03 2025 -0500

    [ci skip] Correct javadoc for Weapon Component (#13096)

commit fba780d6a2948d5a42030a80cf79cb49a8472456
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Tue Oct 7 11:22:54 2025 -0700

    Rebuild patches

    Actually do what the last commit said it was doing...

commit fa57d4b7daea32a825dac7f8c0451e82265bb00b
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Tue Oct 7 11:19:00 2025 -0700

    Remove Vanilla packet processing at start of tick

    Our call to runAllTasksAtTickStart() runs that logic and
    times it correctly, whereas Vanilla does not record task execution
    time for it.

commit 3982efa09840e3dc4795cd4eea37262b7e17d5ac
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Tue Oct 7 11:15:12 2025 -0700

    Sync Moonrise

    See https://github.com/Tuinity/Moonrise/commit/b2968ce710b3365d8b22d537d343477d807bcf45#diff-4ac4a0af7a908348ccdb1ae58ba7819ec41e5ba3f599e058f7060aa06331cacd
    We already perform a watchdog shutdown on Paper so these changes
    are just to keep the codebases in-sync.

commit 8339bb31189ce98b987af35722bc84677b4058cc
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Tue Oct 7 11:04:57 2025 -0700

    Update DataConverter constants for 1.21.10

    https://github.com/PaperMC/DataConverter/commit/77cc8c2f7473a08419734b1ec0b00d13f6f121cb

commit bae47d3e6b11c789a603935162b749db559091e4
Author: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date:   Tue Oct 7 11:00:42 2025 -0700

    Update to 1.21.10 (#13127)

    Co-authored-by: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>

commit 0ff899dae303f986b86ae9c520e477d9a02f5b92
Author: Bjarne Koll <git@lynxplay.dev>
Date:   Tue Oct 7 12:32:34 2025 +0100

    Prevent world mutation on copper golem spawn cancel (#13152)

commit e9fa3a73930d6a8dc10325fd3d75e31bfb1facdd
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Mon Oct 6 20:59:42 2025 -0700

    Use correct queue when blocking on futures

    See https://github.com/Tuinity/Moonrise/commit/a3ac9c7d2e8d264ffa2a1500dcc23ef357ac8c62

commit 914fb084ef60514e74febc0bfd0eadf478d5cd7c
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Mon Oct 6 13:08:03 2025 -0700

    Fix Entity#updateFluidHeightAndDoFluidPushing inconsistency with Vanilla

    See https://github.com/Tuinity/Moonrise/commit/57982e273e560ca4799eefe8247dfee19902852f

commit 8483163895efbe6d71812c1e2e456330c1aa7859
Author: Bjarne Koll <git@lynxplay.dev>
Date:   Mon Oct 6 20:49:10 2025 +0200

    Yield millis from Bukkit#getAverageTickTime

commit b3d215831aa102dd432e25a2eb487877d62337a3
Author: Shane Freeder <theboyetronic@gmail.com>
Date:   Mon Oct 6 18:04:03 2025 +0100

    Disable javadoc.io links

    revert me when jd.io comes back, maybe

commit b4b1f11f40b7b1e0e99c2b67a2647481294f5dea
Author: Pedro <3602279+Doc94@users.noreply.github.com>
Date:   Mon Oct 6 13:55:32 2025 -0300

    [ci skip] Add docs for virtual createMerchant in MerchantInventoryViewBuilder#merchant (#13125)

commit ffd5158a9310ab5c4940e48859c1eee065b06c04
Author: Maddy Miller <git@madelinemiller.dev>
Date:   Tue Oct 7 02:52:40 2025 +1000

    Update 1.21.9 reobf mappings data (#13135)

commit 4dcd8377a2cd24c15ad646814b665abd595455c1
Author: Matt Baxter <matt@phozop.net>
Date:   Sun Oct 5 18:20:51 2025 -0400

    Add CommandSourceStack to UnknownCommandEvent (#13083)

commit 523efa416f0e534564d1339cf08356be09e01014
Author: Bjarne Koll <git@lynxplay.dev>
Date:   Mon Oct 6 00:18:09 2025 +0200

    Flatten currentTick incr into file patches

commit b63dc92e2cb2afdf0f7319b8f8c93999da86a47a
Author: Shane Freeder <theboyetronic@gmail.com>
Date:   Sun Oct 5 18:51:42 2025 +0100

    Restore currentTick counter increment as temp feature patch

commit dc66e8cf6bdea5983b2d4628e6f202d24b3736e4
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Sat Oct 4 19:21:23 2025 -0700

    Return Server#getTPS over minute intervals instead of seconds

commit a4a7461cefdeaa6d230c411e54ae0c26a5a942e3
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Sat Oct 4 11:41:00 2025 -0700

    Rebase fixups

    Oops

commit dcd7847dc0f5c08e7f3f9cd2e77c2e89eae170f0
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Sat Oct 4 11:34:59 2025 -0700

    Improve server tick loop

    See https://github.com/Tuinity/Moonrise/commit/0b842a906ce0cc22bbb3f5c5f228a7dc22a623c4

commit 21b49306abd838773e6db7e55ae566ac33212056
Author: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date:   Fri Oct 3 11:08:41 2025 -0700

    Fix diff in FlowingFluid#canPassThroughWall

    The function should not be returning false if the block states
    are uncachable.

commit e0ba395d37cdc4da0bd0c9061189570aefbb944f
Author: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Date:   Fri Oct 3 10:56:39 2025 -0700

    Add missing plugin ticket comparator fixes #13130

Toffikk committed:
Revert "Rebase for 1.21.10"

This reverts commit 06d3a378ad8daab02e7ed904636553a1ab19de47.

Toffikk committed:
Merge remote-tracking branch 'papermc/main' into better-outdated-version-checking

Toffikk committed:
feat: print the distance to the latest stable release

Toffikk committed:
make this a warn

Toffikk committed:
fixup wording

Toffikk committed:
Merge branch 'PaperMC:main' into better-outdated-version-checking

Toffikk committed:
[ci-skip] Add myself under the MIT license

Toffikk committed:
Merge remote-tracking branch 'papermc/main' into better-outdated-version-checking

Toffikk committed:
Merge remote-tracking branch 'origin/main' into better-outdated-version-checking

Toffikk committed:
Merge remote-tracking branch 'origin/main' into better-outdated-version-checking

kennytv commented on paper-server/patches/sources/net/minecraft/server/MinecraftServer.java.patch:
```diff
@@ -782,6 +782,13 @@
 +            // Paper end - Add onboarding message for initial server start
++            // Paper start - Improve outdated version checking
++            if (System.getProperty("paper.disableStartupVersionCheck") == null) {
++                CompletableFuture.runAsync(() -> {
```
CompletableFutures are scary, this should either at the very least make sure unchecked exceptions aren't swallowed, or use some proper executor

kennytv commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -46,22 +52,68 @@ public long getCacheTime() {
+                default -> {
+                    if (apiResult.isPresent()) {
+                        COMPONENT_LOGGER.error(text("*** You are running an outdated version of Minecraft, which is " + apiResult.get().distance() + " release(s) and " + distance + " build(s) behind!"));
+                        COMPONENT_LOGGER.error(text("*** Please update to the latest stable version on " + DOWNLOAD_PAGE + " ***"));
```
Only use errors for actual errors, a warning is enough

kennytv commented on paper-server/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java:
```diff
@@ -46,22 +52,68 @@ public long getCacheTime() {
+                        COMPONENT_LOGGER.error(text("*** Please update to the latest stable version on " + DOWNLOAD_PAGE + " ***"));
+                    } else {
+                        COMPONENT_LOGGER.warn(text("*** Currently you are " + distance + " build(s) behind ***"));
+                        COMPONENT_LOGGER.warn(text("*** It is highly recommended to download the latest build from " + DOWNLOAD_PAGE + " ***"));
```
Info is enough here, maybe only warn if it exceeds a few weeks like before

Toffikk replied:
thanks for the reviews, i will switch it to use info in this case since imho a time-based check is inherently flawed

Toffikk replied:
ill try and update this pr in like an hour or so

kennytv reviewed (changes requested):
Sorry, somehow we all overlooked that this was time-limited. I didn't read through all the previews reviews, but these things caught my eye

Toffikk committed:
Merge remote-tracking branch 'origin/main' into better-outdated-version-checking

Toffikk committed:
address reviews and cleanup some code

Toffikk committed:
chore: some code cleanup and caching

Toffikk committed:
fix: move connection timeout up and specify a user agent for github

kennytv reviewed (approved):
General approval, though mini still wanted to do or check something

MiniDigger committed:
add config option
